### DISCUSSION TITLE
Recommended structure for small-ish Coalton project?

### DISCUSSION THREAD

**Discussion by ornamentist** [coalton-lang/coalton#364] [Help!] [answered]
I'm new to Common Lisp and Coalton, so this may be a silly question.

Is there an idiomatic way to setup a small-ish Coalton project (let's say it's called "numbers") with ASDF and Common Lisp packages?

 In the project directory root I would have a file `numbers.asd`, containing:

```
(defsystem "numbers"

  ;; various system metadata fields

  :depends-on ("coalton")

  :pathname "src/"

  :serial t

  :components ((:file "package")
               (:file "main")))
```

Then I'd have a `src` directory containing the files `package.lisp` and `main.lisp`. The `package.lisp` file seems to be a convention for defining the packages used in the rest of the source files? It would contain:

```
(in-package #:cl-user)

(defpackage #:numbers
  (:documentation "...")
  (:use #:coalton #:coalton-library)
  (:export ...))
```

And the `main.lisp` file which exports symbols from the `numbers` package would have:

```
(cl:in-package #:numbers)

(coalton-toplevel ;; Coalton specific forms...
  )
```

This is all assuming I've manually installed Coalton somewhere that Quicklisp can find it.

If I didn't want the common `package.lisp` file, but one package per file, what would the directory and `(defpackage...` structure look like?

Any advice much appreciated.

**Comment by stylewarning** [comment, marked as answer]
I recommend actually making two packages, one for CL code and one for Coalton code:

```lisp
(cl:in-package #:cl-user)

(defpackage #:numbers.cl
  (:use #:cl)
  (export ...))

(defpackage #:numbers 
  (:use #:coalton #:coalton-library)
  (:export ...))
```

That way, you can easily write Lisp functions as needed and use them within Coalton, without having to prefix a lot of stuff. Your typical Coalton file would look like

```lisp
(cl:in-package #:numbers.cl)

;; lisp code

(cl:in-package #:numbers)

;; coalton code
```

If you wanted one package per file (a style I've never really appreciated), you'd do the same, just put package definitions at the top of the file. Directory structure would be however you want; Coalton wouldn't require anything.

**Reply by ornamentist** [reply]
Very neat -- thanks for the suggestion.
